antenna/configlib.py

```diff
@@ -0,0 +1,323 @@
+    the ``os.environ``.
+
+    """
+    def get(self, key, namespace=''):
+        if namespace:
+            key = '%s_%s' % (namespace, key)
+
+        key = key.upper()
+        if key in os.environ:
+            return os.environ[key]
+
+        return NO_VALUE
+
+
+class ConfigIniEnv(object):
+    """.ini style configuration layer
+
+    Namespace is a config section. So "foo" in namespace "bar" is::
+
+        [bar]
+        foo=someval
+
+    """
+    def __init__(self, fn):
+        self._parser = ConfigParser()
+        self._parser.readfp(open(fn, 'r'))
+
+    def get(self, key, namespace='main'):
+        if self._parser.has_option(namespace, key):
+            return self._parser.get(namespace, key)
+        return NO_VALUE
+
+
+class ConfigManager(object):
+    """Manages multiple configuration environment layers"""
+    def __init__(self):
+        self.envs = self._initialize_environments()
+
+    def _initialize_environments(self):
+        # Add environments in order of precedence--first is the most important!
+        envs = []
+
+        # Add override environment
+        envs.append(ConfigOverrideEnv())
+
+        # Add OS environment
+        envs.append(ConfigOSEnv())
+
+        # Add .ini environment
+        fn = os.environ.get('ANTENNA_INI', 'antenna.ini')
+        if not os.sep in fn:
+            fn = os.path.join(os.getcwd(), fn)
+        if os.path.exists(fn):
+            envs.append(ConfigIniEnv(fn))
+
+        return envs
+
+    def __call__(self, key, namespace=None, default=NO_VALUE, parser=str,
+                 raise_error=False):
+        """Returns a parsed value from the environment
+
+        :arg key: the key to look up
+        :arg namespace: the namespace for the key--different environments
+            use this differently
+        :arg default: the default value (if any)
+        :arg parser: the parser for converting this value to a Python object
+        :arg raise_error: True if you want a lack of value to raise a
+            ``ConfigurationError``
+
+        Examples::
+
+            # Use the special bool parser
+            DEBUG = config('DEBUG', default=True, parser=bool)
+
+            from antenna.config_util import ListOf
+            ALLOWED_HOSTS = config('ALLOWED_HOSTS', default='localhost',
+                                   parser=ListOf(str))
+
+        """
+        parser = get_parser(parser)
+
+        # Go through environments in reverse order
+        for env in self.envs:
+            val = env.get(key, namespace)
+            if val is not NO_VALUE:
+                return parser(val)
+
+        # Return the default if there is one
+        if default is not NO_VALUE:
+            return parser(default)
+
+        # No value specified and no default, so raise an error to the user
+        if raise_error:
+            raise ConfigurationError(
+                '%s (%s) requires a value of type %s' % (key, namespace, parser)
+            )
```

Comment:
There are two reasons to do this. First is that when calling `.config()` getting back a default value is different than getting back knowledge that the key wasn't set in config at all. With type systems, I really like having a separate "no value was provided" distinct from a default value. There are two cases I was thinking about where this helps.

Say you're writing a component and it uses some config value. Then many moons later, you go to change the component in a way that needs to be backwards compatible, but you want to support new behavior. So maybe you want to check for the new config value and if it's not there, then fall back to the old config value.

The second case is where a component has multiple modes each of which requires different configuration. You see this with components for Django's backends. Some backends require additional configuration that others do not. They try to gloss over this with things like `OPTIONS`, but it's still there. Being able to determine whether a key isn't set at all makes it easier to programmatically know which mode to do.

So, stepping back a bit... Does this configuration management system need this feature? I don't have a need for it in mind, but this kind of thing has bit me in the past and it seemed like a good thing to handle especially where the implementation is pretty straight-forward, easy to test and easy to document.

So, the second thing is whether to make `raise_errors=True` the default. I can't think of which default is better to have: raise errors or not. The `SECRET_KEY` use case is interesting and a compelling use case to make work by default. I'll change it.